protocol/bin-prot/src/de.rs

```diff
@@ -344,6 +355,55 @@ impl<'de, 'a, R: Read> de::Deserializer<'de> for &'a mut Deserializer<R, Strongl
     }
 }
 
+pub struct PolyvarEnum<'a, R: Read, Mode> {
+    de: &'a mut Deserializer<R, Mode>,
+    hash: VariantHash,
+    variants: &'a [&'static str],
+}
+
+impl<'a, 'de, R: Read, Mode> PolyvarEnum<'a, R, Mode> {
+    pub fn new(
+        de: &'a mut Deserializer<R, Mode>,
+        hash: VariantHash,
+        variants: &'a [&'static str],
+    ) -> Self {
+        PolyvarEnum { de, hash, variants }
+    }
+}
+
+macro_rules! impl_polyvar_enum_access {
+    ($ty: ty) => {
+        impl<'de, 'a, R: Read> EnumAccess<'de> for PolyvarEnum<'a, R, $ty> {
+            type Error = Error;
+            type Variant = Enum<'a, R, StronglyTyped>;
+
+            fn variant_seed<V>(self, seed: V) -> Result<(V::Value, Self::Variant)>
+            where
+                V: de::DeserializeSeed<'de>,
+            {
+                // Return the first polyvar variant where the hash matches
+                let index = self
+                    .variants
+                    .into_iter()
+                    .enumerate()
+                    .find_map(|(index, value)| {
+                        if caml_hash_variant(value.as_bytes()) == self.hash {
```

Comment:
Better make sure that value is ascii `[is_ascii](https://doc.rust-lang.org/std/string/struct.String.html#method.is_ascii)(&self) -> [bool](https://doc.rust-lang.org/std/primitive.bool.html)` otherwise computing the caml hash is undefined behaviour. We can throw an error in the case of a non-ascii tag

Reply:
fixed by adding a filter on the iterator `.filter(|v| v.is_ascii())`

Reply:
My issue with this is it will throw the wrong kind of error `UnknownPolyvarTag`. I think it would be better to check and throw an explicit `PolyvarTagNotAscii` error to help with figuring out what is going on